File: .analysis/code-intelligence/scripts/java/Patterns.java
import java.util.*;
import java.util.stream.*;

/**
 * Pattern detection — identifies DI style, error handling, naming conventions.
 */
public class Patterns {

    public record DetectedPatterns(String diStyle, String errorHandling, String naming, String logging, String testing) {}

    public static DetectedPatterns detect(List<String> classes, List<String> functions, List<String> imports) {
        String impText = String.join(" ", imports);
        String classText = String.join(" ", classes);
        return new DetectedPatterns(
            detectDi(impText, functions),
            detectErrorHandling(impText, classText),
            detectNaming(classes),
            detectLogging(impText),
            detectTesting(impText)
        );
    }

    public static String inferModulePurpose(String name, List<String> classes) {
        String all = (name + " " + String.join(" ", classes)).toLowerCase();
        String[][] purposes = {
            {"api", "API layer"}, {"controller", "API layer"},
            {"service", "Business logic"}, {"repository", "Data access"},
            {"config", "Configuration"}, {"shared", "Shared utilities"},
            {"common", "Shared utilities"}, {"model", "Domain model"},
            {"domain", "Domain model"}, {"test", "Testing"},
        };
        for (String[] p : purposes) {
            if (all.contains(p[0])) return p[1];
        }
        return "Application module";
    }

    private static String detectDi(String impText, List<String> functions) {
        if (impText.contains("@Inject") || impText.contains("@Autowired")) return "field injection";
        if (functions.contains("constructor") || functions.contains("__init__")) return "constructor injection";
        return "none";
    }

    private static String detectErrorHandling(String impText, String classText) {
        String all = impText + " " + classText;
        if (all.contains("Result") || all.contains("Either")) return "Result type";
        if (all.contains("ExceptionHandler") || all.contains("ControllerAdvice")) return "exception handler";
        if (all.contains("Exception")) return "try-catch";
        return "unknown";
    }

    private static String detectNaming(List<String> classes) {
        String[] suffixes = {"Controller", "Service", "Repository"};
        List<String> found = new ArrayList<>();
        for (String s : suffixes) {
            if (classes.stream().anyMatch(c -> c.endsWith(s))) found.add("*" + s);
        }
        return found.isEmpty() ? "unknown" : String.join(", ", found);
    }

    private static String detectLogging(String impText) {
        if (impText.contains("slf4j") || impText.contains("SLF4J")) return "SLF4J";
        if (impText.contains("log4j")) return "Log4j";
        if (impText.contains("logging")) return "logging";
        return "unknown";
    }

    private static String detectTesting(String impText) {
        if (impText.contains("junit") || impText.contains("org.junit")) return "JUnit";
        if (impText.contains("pytest") || impText.contains("unittest")) return "pytest";
        if (impText.contains("kotest")) return "kotest";
        if (impText.contains("jest")) return "Jest";
        return "unknown";
    }
}

File: mcp-code-intelligence-python/src/mcp_code_intel/patterns.py
"""Pattern detection — identifies DI style, error handling, naming, logging, testing."""

from typing import TypedDict


class DetectedPatterns(TypedDict):
    """Detected coding patterns for a module."""

    di_style: str
    error_handling: str
    naming: str
    logging: str
    testing: str


def detect_patterns(
    classes: list[dict], functions: list[dict], imports: list[str]
) -> DetectedPatterns:
    """Detect coding patterns from aggregated parse results."""
    return {
        "di_style": _detect_di(classes, functions, imports),
        "error_handling": _detect_error_handling(classes, imports),
        "naming": _detect_naming(classes),
        "logging": _detect_logging(imports),
        "testing": _detect_testing(imports),
    }


def infer_module_purpose(
    name: str, classes: list[dict], packages: list[str]
) -> str:
    """Infer a module's purpose from its contents."""
    all_names = " ".join([name] + [c["name"] for c in classes] + packages).lower()
    purposes = [
        ("api", "API layer"), ("controller", "API layer"),
        ("service", "Business logic"), ("business", "Business logic"),
        ("repository", "Data access"), ("dao", "Data access"),
        ("data", "Data access"),
        ("config", "Configuration"), ("configuration", "Configuration"),
        ("common", "Shared utilities"), ("shared", "Shared utilities"),
        ("test", "Testing"), ("spec", "Testing"),
        ("web", "Web/UI layer"), ("ui", "Web/UI layer"),
        ("model", "Domain model"), ("domain", "Domain model"),
    ]
    for keyword, purpose in purposes:
        if keyword in all_names:
            return purpose
    return "Application module"


def _detect_di(
    classes: list[dict], functions: list[dict], imports: list[str]
) -> str:
    """Detect dependency injection style."""
    all_text = " ".join(imports)
    if "@Inject" in all_text or "@Autowired" in all_text:
        return "field injection"
    if any(f["name"] in ("constructor", "__init__") for f in functions):
        return "constructor injection"
    return "none"


def _detect_error_handling(classes: list[dict], imports: list[str]) -> str:
    """Detect error handling approach."""
    names_text = " ".join(imports + [c["name"] for c in classes])
    if "Result" in names_text or "Either" in names_text:
        return "Result type"
    if "ExceptionHandler" in names_text or "ControllerAdvice" in names_text:
        return "exception handler"
    if "Exception" in names_text or "Error" in names_text:
        return "try-catch"
    return "unknown"


def _detect_naming(classes: list[dict]) -> str:
    """Detect naming conventions from class suffixes."""
    suffixes = ["Controller", "Service", "Repository"]
    found = [
        f"*{s}" for s in suffixes
        if any(c["name"].endswith(s) for c in classes)
    ]
    return ", ".join(found) if found else "unknown"


def _detect_logging(imports: list[str]) -> str:
    """Detect logging framework from imports."""
    imp_text = " ".join(imports)
    if "slf4j" in imp_text or "SLF4J" in imp_text:
        return "SLF4J"
    if "log4j" in imp_text or "Log4j" in imp_text:
        return "Log4j"
    if "logging" in imp_text:
        return "logging"
    return "unknown"


def _detect_testing(imports: list[str]) -> str:
    """Detect testing framework from imports."""
    imp_text = " ".join(imports)
    if "junit" in imp_text or "org.junit" in imp_text:
        return "JUnit"
    if "pytest" in imp_text or "unittest" in imp_text:
        return "pytest"
    if "jest" in imp_text:
        return "Jest"
    if "kotest" in imp_text:
        return "kotest"
    if "vitest" in imp_text:
        return "vitest"
    return "unknown"
